geo/src/algorithm/euclidean_distance.rs

```diff
@@ -480,47 +397,248 @@ where
+                      Geometry::Polygon(p) => p.euclidean_distance(other),
+                      Geometry::MultiPoint(mp) => mp.euclidean_distance(other),
+                      Geometry::MultiLineString(mls) => mls.euclidean_distance(other),
+                      Geometry::MultiPolygon(mp) => mp.euclidean_distance(other),
+                      Geometry::GeometryCollection(gc) => gc.euclidean_distance(other),
+                      Geometry::Rect(r) => r.euclidean_distance(other),
+                      Geometry::Triangle(t) => t.euclidean_distance(other),
+                  }
+              }
+          }
+      )*
+  };
+}
+
+impl_euclidean_distance_to_geometry!([Point<T>, Line<T>, LineString<T>, Polygon<T>, MultiPolygon<T>, Triangle<T>, Rect<T>, GeometryCollection<T>]);
+
+/// Euclidean distance implementation for multi geometry types.
+macro_rules! impl_euclidean_distance_for_iter_geometry {
+  ($for:ty,  [$($target:ty),*]) => {
+      $(
+          impl<T> EuclideanDistance<T, $target> for $for
+          where
+              T: GeoFloat + FloatConst + RTreeNum + FloatConst,
+          {
+              fn euclidean_distance(&self, target: &$target) -> T {
+                  self
+                    .iter()
+                    .map(|g| g.euclidean_distance(target))
+                    .fold(<T as Bounded>::max_value(), |accum, val| accum.min(val))
+              }
+          }
+
+          impl<T> EuclideanDistance<T, $for> for $target
+          where
+              T: GeoFloat + FloatConst + RTreeNum + FloatConst,
+          {
+              fn euclidean_distance(&self, other: &$for) -> T {
+                  other.euclidean_distance(self)
+              }
+          }
+      )*
+  };
+  ($for:ty) => {
+    impl<T> EuclideanDistance<T, $for> for $for
+    where
+        T: GeoFloat + FloatConst + RTreeNum + FloatConst,
+    {
+        fn euclidean_distance(&self, other: &$for) -> T {
+            self
+              .iter()
+              .map(|g| g.euclidean_distance(other))
+              .fold(<T as Bounded>::max_value(), |accum, val| accum.min(val))
+        }
     }
+  };
 }
 
-/// MultiPolygon to Line distance
-impl<T> EuclideanDistance<T, Line<T>> for MultiPolygon<T>
-where
-    T: GeoFloat + FloatConst + Signed + RTreeNum,
-{
-    fn euclidean_distance(&self, other: &Line<T>) -> T {
-        other.euclidean_distance(self)
+impl_euclidean_distance_for_iter_geometry!(MultiPoint<T>, [Point<T>, Line<T>, LineString<T>, MultiLineString<T>, Polygon<T>, MultiPolygon<T>, GeometryCollection<T>]);
+impl_euclidean_distance_for_iter_geometry!(MultiLineString<T>, [Point<T>, Line<T>, LineString<T>, Polygon<T>, MultiPolygon<T>, GeometryCollection<T>]);
+impl_euclidean_distance_for_iter_geometry!(MultiPolygon<T>, [Point<T>, Line<T>, LineString<T>, Polygon<T>, GeometryCollection<T>]);
+impl_euclidean_distance_for_iter_geometry!(GeometryCollection<T>, [Point<T>, Line<T>, LineString<T>, Polygon<T>]);
```

Comment:
Each or these macros is taking a smaller list of input for the second parameter. I assume that's because the implementations have already been covered by a previous macro. 

The trouble is, it's hard to review this and be confident that we didn't miss one.

For example, while `geometry.euclidean_distance(&multi_polygon)` is implemented, we are missing `geometry.euclidean_distance(&multi_point)`.

One idea I had is, even though it might be more typing, to not have the macro implementation be reflexive.
e.g. only implement `impl<T> EuclideanDistance<T, $target> for $for`, but not the inverse `impl<T> EuclideanDistance<T, $for> for $target`.

I think that might make it easier to scan the source code for missing combinations. e.g. something like

```
impl_euclidean_distance_for_iter_geometry!(MultiPoint<T>,         [Point<T>, MultiPoint<T>, Line<T>, LineString<T>, MultiLineString<T>, Polygon<T>, MultiPolygon<T>, GeometryCollection<T>, Geometry<T>]);
impl_euclidean_distance_for_iter_geometry!(MultiLineString<T>,    [Point<T>, MultiPoint<T>, Line<T>, LineString<T>, MultiLineString<T>, Polygon<T>, MultiPolygon<T>, GeometryCollection<T>, Geometry<T>]);
impl_euclidean_distance_for_iter_geometry!(MultiPolygon<T>,       [Point<T>, MultiPoint<T>, Line<T>, LineString<T>, MultiLineString<T>, Polygon<T>, MultiPolygon<T>, GeometryCollection<T>, Geometry<T>]);
impl_euclidean_distance_for_iter_geometry!(GeometryCollection<T>, [Point<T>, MultiPoint<T>, Line<T>, LineString<T>, MultiLineString<T>, Polygon<T>, MultiPolygon<T>, GeometryCollection<T>, Geometry<T>]);
```

(Note I included `Geometry` in the second parameter list)

Possibly that second parameter could be baked into a macro itself so you don't have to type it all out, but I don't feel that strongly about that. 

Similarly with the other macros.

Reply:
Thanks, I agree that it's hard to see if any combinations are missing. I've now updated this to make the macro implementations not reflexive, and also re-organized the macros a bit. I hope that it's more readable and clear now.

One note regarding this point:

> geometry.euclidean_distance(&multi_polygon) is implemented, we are missing geometry.euclidean_distance(&multi_point)

In the new version, all the implementations of euclidean distance between `Geometry<T>` and specific geometry types are handled by 2 separate macros, so `Geometry<T>` is not included in the parameter lists for the multi-geometry macros. This seemed cleaner, since we need to have `impl<T> EuclideanDistance<T, Geometry<T>>` for all geometry types.